src/main/java/io/temporal/internal/statemachines/ActivityStateMachine.java

```diff
@@ -0,0 +1,338 @@
+            State.SCHEDULED_EVENT_RECORDED)
+        .add(
+            State.SCHEDULE_COMMAND_CREATED,
+            Action.CANCEL,
+            State.CANCELED,
+            ActivityStateMachine::cancelScheduleCommand)
+        .add(
+            State.SCHEDULED_EVENT_RECORDED,
+            EventType.EVENT_TYPE_ACTIVITY_TASK_STARTED,
+            State.STARTED)
+        .add(
+            State.SCHEDULED_EVENT_RECORDED,
+            EventType.EVENT_TYPE_ACTIVITY_TASK_TIMED_OUT,
+            State.TIMED_OUT,
+            ActivityStateMachine::notifyTimedOut)
+        .add(
+            State.STARTED,
+            EventType.EVENT_TYPE_ACTIVITY_TASK_COMPLETED,
+            State.COMPLETED,
+            ActivityStateMachine::notifyCompleted)
+        .add(
+            State.STARTED,
+            EventType.EVENT_TYPE_ACTIVITY_TASK_FAILED,
+            State.FAILED,
+            ActivityStateMachine::notifyFailed)
+        .add(
+            State.STARTED,
+            EventType.EVENT_TYPE_ACTIVITY_TASK_TIMED_OUT,
+            State.TIMED_OUT,
+            ActivityStateMachine::notifyTimedOut)
+        .add(
+            State.SCHEDULED_EVENT_RECORDED,
+            Action.CANCEL,
+            State.SCHEDULED_ACTIVITY_CANCEL_COMMAND_CREATED,
+            ActivityStateMachine::createRequestCancelActivityTaskCommand)
+        .add(
+            State.SCHEDULED_ACTIVITY_CANCEL_COMMAND_CREATED,
+            EventType.EVENT_TYPE_ACTIVITY_TASK_CANCEL_REQUESTED,
+            State.SCHEDULED_ACTIVITY_CANCEL_EVENT_RECORDED,
+            ActivityStateMachine::notifyCanceledIfTryCancel)
+        .add(
+            State.SCHEDULED_ACTIVITY_CANCEL_COMMAND_CREATED,
+            CommandType.COMMAND_TYPE_REQUEST_CANCEL_ACTIVITY_TASK,
+            State.SCHEDULED_ACTIVITY_CANCEL_COMMAND_CREATED)
+        .add(
+            State.SCHEDULED_ACTIVITY_CANCEL_COMMAND_CREATED,
+            EventType.EVENT_TYPE_ACTIVITY_TASK_STARTED,
+            State.STARTED_ACTIVITY_CANCEL_COMMAND_CREATED)
+        .add(
+            State.SCHEDULED_ACTIVITY_CANCEL_EVENT_RECORDED,
+            EventType.EVENT_TYPE_ACTIVITY_TASK_CANCELED,
+            State.CANCELED,
+            ActivityStateMachine::notifyCanceled)
+        .add(
+            State.SCHEDULED_ACTIVITY_CANCEL_EVENT_RECORDED,
+            EventType.EVENT_TYPE_ACTIVITY_TASK_STARTED,
+            State.STARTED_ACTIVITY_CANCEL_EVENT_RECORDED)
+        .add(
+            State.STARTED_ACTIVITY_CANCEL_COMMAND_CREATED,
+            EventType.EVENT_TYPE_ACTIVITY_TASK_CANCEL_REQUESTED,
+            State.STARTED_ACTIVITY_CANCEL_EVENT_RECORDED,
+            ActivityStateMachine::notifyCanceledIfTryCancel)
+        .add(
+            State.STARTED,
+            Action.CANCEL,
+            State.STARTED_ACTIVITY_CANCEL_COMMAND_CREATED,
+            ActivityStateMachine::createRequestCancelActivityTaskCommand)
+        .add(
+            State.STARTED_ACTIVITY_CANCEL_COMMAND_CREATED,
```

Comment:
for this state and the cancel_command_Created state, we don't have transitions for a timed out event?

Reply:
I've added unit tests that demonstrate that these transitions are not possible. The basic idea is that the workflow code always executes in the context of an event loop. It causes activity timeouts to be delivered before the code that causes the timeout which generates the cancel command.